src/background/sidebar-injector.ts

```diff
@@ -0,0 +1,485 @@
+import { chromeAPI, executeFunction, executeScript } from './chrome-api';
+import { detectContentType } from './detect-content-type';
+import {
+  AlreadyInjectedError,
+  BlockedSiteError,
+  LocalFileError,
+  NoFileAccessError,
+  RestrictedProtocolError,
+} from './errors';
+
+const CONTENT_TYPE_HTML = 'HTML';
+const CONTENT_TYPE_PDF = 'PDF';
+const CONTENT_TYPE_VITALSOURCE = 'VITALSOURCE';
+const CONTENT_TYPE_LMS = 'LMS';
+
+/* istanbul ignore next - Code coverage breaks `eval`-ing of this function in tests. */
+function setClientConfig(config: object) {
+  const script = document.createElement('script');
+  script.className = 'js-hypothesis-config';
+  script.type = 'application/json';
+  script.textContent = JSON.stringify(config);
+  // This ensures the client removes the script when the extension is deactivated
+  script.setAttribute('data-remove-on-unload', '');
+  document.head.appendChild(script);
+}
+
+/**
+ * Function that is run in a frame to test whether the Hypothesis client is
+ * active there.
+ *
+ * @param extensionURL - Root URL for the extension, of the form
+ *   "chrome-extension://{ID}/".
+ */
+function isClientActive(extensionURL: string) {
+  const annotatorLink = document.querySelector(
+    'link[type="application/annotator+html"]'
+  ) as HTMLLinkElement | null;
+  return annotatorLink?.href.startsWith(extensionURL) ?? false;
+}
+
+/**
+ * A Chrome tab for which we have ID and URL information.
+ *
+ * This type avoids the need to check everywhere we access these properties.
+ */
+type Tab = chrome.tabs.Tab & { id: number; url: string };
+
+/**
+ * Check that a tab has the necessary metadata to inject or un-inject the client.
+ *
+ * All "normal" tabs should have this information because of the extension's
+ * permissions.
+ */
+function checkTab(tab: chrome.tabs.Tab): Tab {
+  if (!tab.id || !tab.url) {
+    throw new Error('Tab is missing ID or URL');
+  }
+  return tab as Tab;
+}
```

Comment:
Perhaps we could use a TypeScript assertion here (like a guard, but it throws instead of returning `false`).

```suggestion
function assertTab(tab: chrome.tabs.Tab): asserts tab is Tab {
  if (!tab.id || !tab.url) {
    throw new Error('Tab is missing ID or URL');
  }
}

function checkTab(tab: chrome.tabs.Tab): Tab {
  assertTab(tab);
  return tab;
}
```

This allows to remove the `as Tab` casting, as the assertion tells TS it has the proper type after that check.

But it's more verbose, so I'll leave it up to you.